apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/TimedBufferingConsumer.cs

```diff
@@ -72,44 +68,19 @@ public override void Dispose()
         }
 
         /// <inheritdoc />
-        protected override void ReceiveWithSemaphoreHeld(IEnumerable<T> items)
-        {
-            GaxPreconditions.CheckNotNull(items, nameof(items));
-            _items.AddRange(items);
-        }
-
-        /// <inheritdoc />
-        protected override Task ReceiveAsyncWithSemaphoreHeldAsync(
-            IEnumerable<T> items, CancellationToken cancellationToken = default)
+        protected override bool ReceiveWithSemaphoreHeld(IEnumerable<T> items)
         {
             GaxPreconditions.CheckNotNull(items, nameof(items));
             _items.AddRange(items);
-            return CommonUtils.CompletedTask;
+            return false;
         }
 
         /// <inheritdoc />
-        protected override void FlushWithSemaphoreHeld()
+        protected override IEnumerable<T> GetAndResetItemsWithSemaphoreHeld()
         {
-            if (_items.Count == 0)
-            {
-                return;
-            }
-
-            _consumer.Receive(_items);
-            _items = new List<T>();
-        }
-
-        /// <inheritdoc />
-        protected override Task FlushAsyncWithSemaphoreHeldAsync(CancellationToken cancellationToken)
-        {
-            if (_items.Count == 0)
-            {
-                return CommonUtils.CompletedTask;
-            }
-
             IList<T> old = _items;
```

Comment:
Note that with an `Interlocked.Exchange`, you might be able to do this without needing the semaphore at all.

Reply:
Maybe in the `TimedBufferingConsumer` implementation it's possible, but remember that we are inheriting from `FlushableConsumerBase`, which needs to be general and grab the same semaphore for receiving items and for getting them for flushing. Actually `SizedBufferingConsumer`, also inhereting from `FlushableConsumerBase` does a little bit more in its implementation of this method, that is setting the buffer size to zero.